Translation of the “Updating Arrays in State” page (#542)
reactjs/fr.react.dev: src/content/learn/updating-arrays-in-state.md

</Sandpack>

Ici, vous utilisez d'abord la syntaxe de *spread* `[...list]` pour créer une copie du tableau d'origine. Maintenant que vous avez une copie, vous pouvez utiliser des méthodes modifiantes comme `nextList.reverse()` ou `nextList.sort()`, ou même affecter individuellement des éléments avec `nextList[0] = "quelque chose"`.

> Cependant, **même si vous copiez un tableau, vous ne pouvez pas modifier directement les éléments existants _à l'intérieur_ de celui-ci.** C'est parce que la copie est superficielle : le nouveau tableau contiendra les mêmes éléments que le tableau d'origine. Ainsi, si vous modifiez un objet à l'intérieur du tableau copié, vous modifierez l'état existant. Par exemple, le code suivant est problématique.

Reviewer: ```suggestion
Cependant, **même si vous copiez un tableau, vous ne pouvez pas modifier directement les éléments existants _à l'intérieur_ de celui-ci.** C'est parce que la copie est superficielle : le nouveau tableau contiendra les mêmes éléments que le tableau d'origine. Ainsi, si vous modifiez un objet à l'intérieur du tableau copié, vous modifiez l'état existant. Par exemple, le code suivant est problématique.
```

## Mettre à jour des objets dans des tableaux {/*updating-objects-inside-arrays*/}

Les objets ne sont pas _vraiment_ « à l'intérieur » des tableaux. Ils peuvent sembler être « à l'intérieur » dans le code, mais chaque objet dans un tableau est une valeur distincte vers laquelle le tableau « pointe ». C'est pourquoi vous devez faire attention lorsque vous modifiez des champs imbriqués tels que `list[0]`. La liste d'œuvres d'art d'une autre personne peut pointer vers le même élément du tableau !

**Lorsque vous mettez à jour un état imbriqué, vous devez créer des copies à partir de l'endroit où vous souhaitez effectuer la mise à jour, en remontant jusqu’au plus haut niveau.** Voyons comment cela fonctionne.

> Dans cet exemple, deux listes d'œuvres d'art séparées ont le même état initial. Elles sont censées être isolées, mais à cause d'une mutation, leur état est accidentellement partagé, et cocher une case dans l'une des listes affecte l'autre liste :

Reviewer: ```suggestion
Dans cet exemple, deux listes d'œuvres d'art séparées ont le même état initial. Elles sont censées être isolées, mais à cause d'une modification directe, leur état est accidentellement partagé, et cocher une case dans l'une des listes affecte l'autre liste :
```

</Sandpack>

> Le problème se trouve dans un code comme celui-ci :

Reviewer: ```suggestion
Le problème se trouve dans du code comme celui-ci :
```

```js
const myNextList = [...myList];
const artwork = myNextList.find(a => a.id === artworkId);
artwork.seen = nextSeen; // Problème : modifie un élément existant
setMyList(myNextList);
```

Bien que le tableau `myNextList` soit nouveau, les *éléments eux-mêmes* sont les mêmes que dans le tableau `myList` d'origine. Donc, en changeant `artwork.seen`, vous modifiez l'œuvre d'art *d'origine*. Cette œuvre d'art est également dans `yourList`, ce qui provoque le bug. Des bugs comme celui-ci peuvent être difficiles à comprendre, mais heureusement, ils n'ont pas lieu si vous évitez de modifier l'état.

**Vous pouvez utiliser `map` pour remplacer un ancien élément par sa nouvelle version sans mutation.**

```js
setMyList(myList.map(artwork => {
  if (artwork.id === artworkId) {

>     // Créez un *nouvel* objet avec les modifications

Reviewer: ```suggestion
    // Crée un *nouvel* objet avec les modifications
```

    return { ...artwork, seen: nextSeen };
  } else {
    // Pas de changement
    return artwork;
  }
}));
```

> Ici, `...` est la syntaxe de *spread* d'objets utilisée pour [créer une copie d'un objet.](/learn/updating-objects-in-state#copying-objects-with-the-spread-syntax)

Reviewer: ```suggestion
Ici, `...` est la syntaxe de *spread* d'objets utilisée pour [créer une copie d'un objet](/learn/updating-objects-in-state#copying-objects-with-the-spread-syntax).
```

> Avec cette approche, aucun des éléments de l'état n'est modifié et le bug est corrigé :

Reviewer: ```suggestion
Avec cette approche, aucun des éléments de l'état existant n'est modifié et le bug est corrigé :
```

  function handleToggleMyList(artworkId, nextSeen) {
    setMyList(myList.map(artwork => {
      if (artwork.id === artworkId) {

>         // Créez un *nouvel* objet avec les modifications

Reviewer: ```suggestion
        // Crée un *nouvel* objet avec les modifications
```

  function handleToggleYourList(artworkId, nextSeen) {
    setYourList(yourList.map(artwork => {
      if (artwork.id === artworkId) {

>         // Créez un *nouvel* objet avec les modifications

Reviewer: ```suggestion
        // Crée un *nouvel* objet avec les modifications
```

</Sandpack>

> En général, **vous ne devriez modifier que les objets que vous venez de créer.** Si vous insérez une *nouvelle* œuvre d'art, vous pouvez la modifier, mais si vous traitez quelque chose qui est déjà dans l'état, vous devez faire une copie.

Reviewer: ```suggestion
En général, **vous ne devriez modifier que les objets que vous venez de créer**. Si vous insérez une *nouvelle* œuvre d'art, vous pouvez la modifier, mais si vous traitez quelque chose qui est déjà dans l'état, vous devez faire une copie.
```

### Écrire une logique de mise à jour concise avec Immer {/*write-concise-update-logic-with-immer*/}

> Mettre à jour des tableaux imbriqués sans mutation peut conduire à du code un peu répétitif. [Tout comme avec les objets](/learn/updating-objects-in-state#write-concise-update-logic-with-immer):

Reviewer: ```suggestion
Mettre à jour des tableaux imbriqués sans modification directe peut conduire à du code un peu répétitif. [Tout comme avec les objets](/learn/updating-objects-in-state#write-concise-update-logic-with-immer) :
```

});
```

C'est parce que vous ne modifiez pas l'état _d'origine_, mais un objet `draft` spécial fourni par Immer. De même, vous pouvez appliquer des méthodes modifiantes telles que `push()` et `pop()` au contenu du `draft`.

> En interne, Immer construit toujours le prochain état à partir de zéro en fonction des changements que vous avez apportés au `draft`. Cela permet de maintenir vos gestionnaires d'événements très concis sans jamais modifier l'état.

Reviewer: ```suggestion
En interne, Immer construit toujours le prochain état à partir de zéro en fonction des changements que vous avez apportés au `draft`. Ça permet de garder des gestionnaires d'événements très concis sans jamais modifier l'état directement.
```

<Solution>

> Avec Immer, vous pouvez écrire du code de manière modifiante, tant que vous ne modifiez que les parties du `draft` que Immer vous fournit. Ici, toutes les mutations sont effectuées sur le `draft` de sorte que le code fonctionne correctement:

Reviewer: ```suggestion
Avec Immer, vous pouvez écrire du code de manière modifiante, tant que vous ne modifiez que des parties du `draft` que Immer vous fournit. Ici, toutes les modifications sont effectuées sur le `draft` de sorte que le code fonctionne correctement :
```
